src/main/java/com/google/devtools/build/lib/bazel/bzlmod/VendorUtil.java

```diff
@@ -0,0 +1,162 @@
+      if (!isRepoUpToDate(repo.getName(), externalRepoRoot)) {
+        Path repoUnderVendor = vendorDirectory.getRelative(repo.getName());
+        if (!repoUnderVendor.exists()) {
+          repoUnderVendor.createDirectory();
+        }
+        FileSystemUtils.copyTreesBelow(
+            externalRepoRoot.getRelative(repo.getName()), repoUnderVendor, Symlinks.NOFOLLOW);
+        FileSystemUtils.copyFile(
+            externalRepoRoot.getChild("@" + repo.getName() + ".marker"),
+            vendorDirectory.getChild("@" + repo.getName() + ".marker"));
+      }
+    }
+  }
+
+  /**
+   * Checks if the given URL is vendored.
+   *
+   * @param url The URL to check.
+   * @return true if the URL is vendored, false otherwise.
+   * @throws UnsupportedEncodingException if the URL decoding fails.
+   */
+  public boolean isUrlVendored(URL url) throws UnsupportedEncodingException {
+    return getVendorPathForURL(url).isFile();
+  }
+
+  /**
+   * Vendors the registry URL with the specified content.
+   *
+   * @param url The registry URL to vendor.
+   * @param content The content to write.
+   * @throws IOException if an I/O error occurs.
+   */
+  public void vendorRegistryURL(URL url, byte[] content)
+      throws IOException {
+    Path outputPath = getVendorPathForURL(url);
+    Objects.requireNonNull(outputPath.getParentDirectory()).createDirectoryAndParents();
+    FileSystemUtils.writeContent(outputPath, content);
+  }
+
+  /**
+   * Reads the content of the registry URL and verifies its checksum.
+   *
+   * @param url The registry URL to read.
+   * @param checksum The checksum to verify.
+   * @return The content of the registry URL.
+   * @throws IOException if an I/O error occurs or the checksum verification fails.
+   */
+  public byte[] readRegistryURL(URL url, Checksum checksum)
+      throws IOException {
+    byte[] content = FileSystemUtils.readContent(getVendorPathForURL(url));
+    Hasher hasher = checksum.getKeyType().newHasher();
+    hasher.putBytes(content);
+    HashCode actual = hasher.hash();
+    if (!checksum.getHashCode().equals(actual)) {
+      throw new IOException(
+          String.format(
+              "Checksum was %s but wanted %s",
+              checksum.emitOtherHashInSameFormat(actual),
+              checksum.emitOtherHashInSameFormat(checksum.getHashCode())));
+    }
+    return content;
+  }
+
+  /**
+   * Checks if the repository under vendor dir needs to be updated by comparing its marker file with the
+   * one under <output_base>/external. This function assumes the marker file under <output_base>/external exists
+   * and is up-to-date.
+   *
+   * @param repoName The name of the repository.
+   * @param externalPath The root directory of the external repositories.
+   * @return true if the repository is up-to-date, false otherwise.
+   * @throws IOException if an I/O error occurs.
+   */
+  private boolean isRepoUpToDate(String repoName, Path externalPath)
```

Comment:
Could this take in a `RepositoryName`? Then we clearly wouldn't have to worry about slashes in `repoName`.

Reply:
SG, also extracted the marker file name to a function in `RepositoryName`